torch/csrc/cuda/nccl.cpp

```diff
@@ -0,0 +1,175 @@
+#include "nccl.h"
+#include "torch/csrc/THP.h"
+#include "torch/csrc/Types.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/cuda/THCP.h"
+
+#include <nccl.h>
+#include <sstream>
+#include <unordered_map>
+
+static inline void CHECK(ncclResult_t status)
+{
+  if (status != ncclSuccess) {
+    std::stringstream err;
+    err << "NCCL Error " << status << ": " << ncclGetErrorString(status);
+    throw std::runtime_error(err.str());
+  }
+}
+
+// TODO: make this thread_local + add mutexes
+std::unordered_map<std::string, ncclComm_t*> _communicators;
+
+PyObject * THCPModule_nccl_destroy(PyObject *args) {
+  for(auto it: _communicators) {
+    ncclCommDestroy(*(it.second));
+  }
+  Py_RETURN_NONE;
+}
+
+static ncclComm_t* _get_communicator(std::vector<at::Tensor>& inputs, int ndevices) {
+  std::stringstream hash;
+  std::vector<int> devs;
+  for (int i = 0; i < ndevices; i++) {
+    int dev = inputs[i].get_device();
+    hash <<  dev << ",";
+    devs.push_back(dev);
+  }
+  if (_communicators.find(hash.str()) == _communicators.end()) {
+    ncclComm_t *comms = (ncclComm_t*) malloc(sizeof(ncclComm_t) * ndevices);
+    if (comms == NULL) throw std::runtime_error("failed to allocate ncclComm_t*");
+    CHECK(ncclCommInitAll(comms, ndevices, &devs[0]));
+    _communicators[hash.str()] = comms;
+  }
+  return _communicators[hash.str()];
```

Comment:
this if/return can be implemented more efficiently like this (written assuming you already changed unordered map keys to `at::IntList`):
```cpp
auto it = _communicators.find(devices);
if (it == _communicators.end()) {
  // alloc comunicator
  std::tie(it, std::ignore) = _communicators.emplace(devices, comms);
}
return *it;
```

Reply:
this logic looks pretty convoluted. i think i prefer the code as it currently exists.

Reply:
It's a standard C++ idiom and it's more efficient. NCCL functions are pretty important so I'd keep them performant